File: python/collatz.py
#!/usr/bin/python

def countCollatz(n):
    count = 0        # starts at 1 since the 1 at the end of the sequence won't be counted
    while (n != 1):
        if (n%2 == 1):
            n = 3*n + 1
        else: 
            n = n / 2
        count = count + 1

    return count

num = []    # parallel arrays used to hold integer n and the collatz sequence length
length = []
# zero the list
for i in range(10):
    num.append(0)
    length.append(0)

n = 1               # value to compute the length of the collatz sequence for
max_value = 5000000000   # collatz values to compute length of
for i in xrange(max_value):
    result = countCollatz(n)    # length of collatz sequence for n
    if (result > length[9]):    # determine if result is a max
        offset = 0
        while (result < length[offset]):    # determine where result should go
            offset = offset + 1

        # adjust the positions of the other values in lists
        temp = offset
        offset = 9
        while (offset > temp):
            num[offset] = num[offset - 1]
            length[offset] = length[offset - 1]
            offset = offset - 1
        # put numbers in list
        num[temp] = n
        length[temp] = result    
    n = n + 1                   # increment the integer to be computed

# print the list
print("The 10 longest Collatz Sequence lengths sorted by length: ")
for i in range (9): # print the list of the 10 largest collatz seqences
    print("n: %s length: %s" % (num[i], length[i]))

File: perl/collatz.pl
#!/usr/bin/perl
use warnings;

# function that returns the length of the collatz sequence of given n
sub CountCollatz      
{
    my $n = $_[0];    # special array holds function arguments in perl, this accesses the one element passed and puts it in n
    my $count = 0;

    while ($n != 1)
    {
        if ($n%2 == 1)  # if odd
        {
            $n = 3*$n + 1;
        }
        else            # if even
        {
            $n = $n / 2;
        }
        $count = $count + 1;
    }
    return $count;
}

# now we'll start our main...

# we will use parallel arrays nums and lengths to hold the value for which the collatz
# sequence length should be computed for and the collatz length, respectively
for ($i = 0; $i < 10; $i = $i + 1)
{
    $nums[$i] = 0;       # zero the arrays
    $vals[$i] = 0;
}

$max_value = 5000000000;
$n = 1;

for ($j = 0; $j < $max_value; $j = $j + 1)
{
    $result = CountCollatz($n);                     # compute collatz length of $n
    if ($result > $vals[9])                         # check if it is a new max
    {
        $offset = 0;
        while ($result < $vals[$offset])            # find where it should go in the list
        {
            $offset = $offset + 1;
        }

        $temp = $offset;
        $offset = 9;
        while ($offset > $temp)                     # shift the rest of the values back into place
        {
            $nums[$offset] = $nums[$offset - 1];
            $vals[$offset] = $vals[$offset - 1];
            $offset = $offset - 1;
        }
        $nums[$temp] = $n;                          # store $n and collatz length in nums
        $vals[$temp] = $result;                     # and vals
    }
    $n = $n + 1;                                    # increment $n
}

# print the things in order of collat sequence length
print"The 10 integers with the largest collatz sequence lengths are: \n";
for ($k = 0; $k < 10; $k = $k + 1)
{
    print"$nums[$k]  $vals[$k]\n";
}